### DISCUSSION TITLE
Add command for merging into another branch without removing the current branch

### DISCUSSION THREAD

**Discussion by bryan-lott** [magit/magit#4597] [Ideas]
A common workflow for myself and I suspect others as well is to merge from the current branch into another branch, push that branch to remote for testing/click-testing/PR's/etc and continue development on the existing current branch.

Example of current workflow:
1. Current branch: `feature-branch-1`
2. Checkout branch: `stage` -> `b c stage`
3. Pull `stage` from remote -> `F p`
4. Merge `feature-branch-1` into `stage` -> `m m feature-branch-1`
5. Push `stage` to remote -> `p p`
6. Checkout `feature-branch-1` to continue development `b c feature-branch-1`

Each of those commands can take up to several seconds to complete meaning that the entire procedure takes time and is error prone.

Example of proposed workflow:
1. Current branch: `feature-branch-1`
2. Pull `stage` from remote -> not sure if there is an existing command for this, in a perfect world this would be combined with the next step
3. Merge `feature-branch-1` into `stage` -> `m [?] stage`
4. Push `stage` to remote -> `p o stage`

This is similar to `magit-merge-into` _without_ removing the current branch, implying that a flag on merging may be a better route?

Happy to collaborate and refine the above if it would help.  My elisp skills are rudimentary at best and my git porcelain skills are nonexistent.  The idea comes from a workflow I've been using in GitKraken but I'd love to stay inside emacs if at all possible.

edit: just realized I missed a step in pulling stage from the remote, added it into existing and proposed workflows

**Comment by tarsius** [comment]
I never do that and it is not something I would encourage.

Let's wait and see if others also feel this would be a useful addition.

**Reply by bryan-lott** [reply]
Unfortunately this has been a common pattern at a number of companies I've worked for.  `stage` (i.e., the environment below production) is used instead of testing locally.  Happy to wait and see if others have also run into the same pattern.

**Reply by bard** [reply]
This is common in push-to-deploy scenarios. The instance I bump into most often looks like this:

1. switch from `develop` branch to `release` branch
2. merge `develop` ~~onto~~ into `release`
3. push `release`
4. forget that I'm on `release`
5. keep working on and committing to `release` 
6. at some point, realize and fix

That said, it does look a bit too specialized for a magit command. `magit-merge-into-and-push` starts sounding like the domain of macros rather than commands.

**Reply by tarsius** [reply]
> 2. merge `develop` onto `release`

merge `develop` **into** `release`
or
**rebase** `develop` onto `release`
?

That's a first problem a `checkout+merge+push+checkout` command would encounter. Workflows differ, so we probably would need two variants.

Also, regardless of whether merge or rebase is used, 2 will likely fail, requiring that the user addresses conflicts. In that case `checkout+merge+push+checkout` could just give up and let the user deal with 3 themselves. That could result in that command being perceived as "unreliable".

Alternative we could go all in make this as complex as magit's support for rebasing itself, allowing the user to finish or abort the interrupted workflow. That would be a lot of work and potentially confusing. So I don't think this would be worth the effort.

**Reply by bard** [reply]
I used the wrong preposition, I meant "merge into" (both `develop` and `release` are public and long lived).

I'm not sure why 2. would _likely_ fail? `release` is never committed to, only merged into from `develop`.

That said, your point still stands, because when it comes to the `push` stage, the origin's `release` might already have been updated by someone else, so the `push` would fail, and handling it within the magit flow is probably more effort than it's worth.

**Reply by tarsius** [reply]
> That said, it does look a bit too specialized for a magit command.

Yes. There are too many other ideas, and I cannot implement everything.
